File: src/display/renderer/mesh/mesh.rs
//
// Created by Sinis on 08/05/2025.
//

use gl::types::{GLsizei, GLuint};
use glutin_winit::finalize_window;
use crate::display::renderer::mesh::shader::shader::Shader;
use crate::display::renderer::mesh::texture::texture::Texture;
use crate::display::renderer::mesh::vertex::vertex::Vertex;

pub struct Mesh {
    vao : GLuint,
    vbo : GLuint,
    ebo : GLuint,
    vertices : Vec<Vertex>,
    indices : Vec<u32>,
    textures: Vec<Texture>
}

impl Mesh {
    pub unsafe fn new(vertices: Vec<Vertex>, indices: Vec<u32>, textures: Vec<Texture>) -> Mesh {
        let (vao, vbo, ebo) = (0, 0, 0);
        let mut mesh = Self {
            vao,
            vbo,
            ebo,
            vertices,
            indices,
            textures
        };
        mesh.setup_mesh();
        mesh
    }

    pub unsafe fn setup_mesh(&mut self) {
        gl::CreateBuffers(1, &mut self.vbo);
        gl::NamedBufferData(self.vbo, (self.vertices.len() * std::mem::size_of::<Vertex>()).cast_signed(), self.vertices.as_ptr() as *const _, gl::STATIC_DRAW);

        gl::CreateBuffers(1, &mut self.ebo);
        gl::NamedBufferData(self.ebo, (self.indices.len() * std::mem::size_of::<u32>()).cast_signed(), self.indices.as_ptr() as *const _, gl::STATIC_DRAW);

        gl::CreateVertexArrays(1, &mut self.vao);
        gl::VertexArrayVertexBuffer(self.vao, 0, self.vbo, 0, std::mem::size_of::<Vertex>() as i32);
        gl::VertexArrayElementBuffer(self.vao, self.ebo);

        //Enable vertex attributes (location = ?)
        gl::EnableVertexArrayAttrib(self.vao, 0);
        gl::EnableVertexArrayAttrib(self.vao, 1);
        gl::EnableVertexArrayAttrib(self.vao, 2);

        gl::VertexArrayAttribFormat(self.vao, 0, 3, gl::FLOAT, gl::FALSE, core::mem::offset_of!(Vertex, position) as u32);
        gl::VertexArrayAttribFormat(self.vao, 1, 3, gl::FLOAT, gl::FALSE, core::mem::offset_of!(Vertex, normal) as u32);
        gl::VertexArrayAttribFormat(self.vao, 2, 2, gl::FLOAT, gl::FALSE, core::mem::offset_of!(Vertex, tex_coords) as u32);

        gl::VertexArrayAttribBinding(self.vao, 0, 0);
        gl::VertexArrayAttribBinding(self.vao, 1, 0);
        gl::VertexArrayAttribBinding(self.vao, 2, 0);

        // Logs::debug("Mesh created with self.vbo: " + std::to_string(self.vbo) + ", self.ebo: " + std::to_string(self.ebo) + ", self.vao: " + std::to_string(self.vao));
        // check_opengl::error("mesh");
    }

    pub unsafe fn draw(&mut self, shader: &Shader) {     // render the mesh
        // bind appropriate textures
        let diffuse_nr = 1;
        let specular_nr = 1;
        let normal_nr = 1;
        let height_nr = 1;
        for i in 0..self.textures.len() {
            gl::ActiveTexture(self.textures.get(i).unwrap().get_code()); // active proper texture unit before binding
            // retrieve texture number (the N in diffuse_textureN)
            let number: String;
            let name = self.textures[i].get_t_type();
            if (name == "texture_diffuse") {
                number = (diffuse_nr + 1).to_string();
            } else if (name == "texture_specular") {
                number = (specular_nr + 1).to_string(); // transfer unsigned int to string
            } else if (name == "texture_normal") {
                number = (normal_nr + 1).to_string(); // transfer unsigned int to string
            } else if (name == "texture_height") {
                number = (height_nr + 1).to_string(); // transfer unsigned int to string
            }

            // now set the sampler to the correct texture unit
            shader.set_int(name.as_str(), i as i32);
            // and finally bind the texture
            gl::BindTexture(gl::TEXTURE_2D, self.textures[i].get_texture());
        }

        // draw mesh
        gl::BindVertexArray(self.vao);
        gl::DrawElementsBaseVertex(gl::TRIANGLES, self.indices.len() as GLsizei, gl::UNSIGNED_INT, std::ptr::null(), 0);
        gl::BindVertexArray(0);


        // always good practice to set everything back to defaults once configured.
        gl::ActiveTexture(gl::TEXTURE0);
    }

    pub unsafe fn load_textures(&mut self, filename : String, t_code: u32, name : String) {
        let texture: Texture = Texture::new(filename, t_code, name).unwrap();
        self.textures.push(texture);
    }

    pub unsafe fn init_textures(&self) {
        // set the texture wrapping/filtering options (on the currently bound texture object)
        gl::TextureParameteri(gl::TEXTURE_2D, gl::TEXTURE_WRAP_S, gl::REPEAT as i32);
        gl::TextureParameteri(gl::TEXTURE_2D, gl::TEXTURE_WRAP_T, gl::REPEAT as i32);

        // Note: You had REPEAT for filters in your snippet.
        // Usually, you want NEAREST or LINEAR for filters.
        gl::TextureParameteri(gl::TEXTURE_2D, gl::TEXTURE_MIN_FILTER, gl::NEAREST as i32);
        gl::TextureParameteri(gl::TEXTURE_2D, gl::TEXTURE_MAG_FILTER, gl::NEAREST as i32);
    }
}
impl Drop for Mesh {
    fn drop(&mut self) {
        unsafe {
            gl::DeleteVertexArrays(1, &self.vao);
            gl::DeleteBuffers(1, &self.vbo);
            gl::DeleteBuffers(1, &self.ebo);
            for texture in self.textures.iter() {
                gl::DeleteTextures(1, &texture.get_texture());
            }
        }
    }
}

File: src/game/render/mesh/mesh.cpp
//
// Created by Sinis on 08/05/2025.
//

#include "mesh.h"
#include "stb_image.h"
#include <utility>
#include <vector>
#include <cstdio>
#include <string>

#include "../../../utils/logs/Logs.h"
#include "glad/glad.h"

Mesh::Mesh(std::vector<VERTEX> vertices, std::vector<unsigned int> indices, std::vector<TEXTURE> textures) {
    this->vertices = std::move(vertices);
    this->indices = std::move(indices);
    this->textures = std::move(textures);
    this->VAO = 0, this->VBO = 0, this->EBO = 0;
    setupMesh();
}

void Mesh::setupMesh() {
    glCreateBuffers(1, &VBO);
    glNamedBufferData(VBO,vertices.size() * sizeof(VERTEX), &vertices[0], GL_STATIC_DRAW);

    glCreateBuffers(1, &EBO);
    glNamedBufferData(EBO, indices.size() * sizeof(unsigned int),&indices[0],GL_STATIC_DRAW);

    glCreateVertexArrays(1, &VAO);
    glVertexArrayVertexBuffer(VAO, 0, VBO, 0, sizeof(VERTEX));
    glVertexArrayElementBuffer(VAO, EBO);

    //Enable vertex attributes (location = ?)
    glEnableVertexArrayAttrib(VAO, 0);
    glEnableVertexArrayAttrib(VAO, 1);
    glEnableVertexArrayAttrib(VAO, 2);

    glVertexArrayAttribFormat(VAO,0,3,GL_FLOAT, GL_FALSE, offsetof(VERTEX, Position));
    glVertexArrayAttribFormat(VAO,1,3,GL_FLOAT, GL_FALSE,offsetof(VERTEX, Normal));
    glVertexArrayAttribFormat(VAO,2,2,GL_FLOAT, GL_FALSE,offsetof(VERTEX, TexCoords));

    glVertexArrayAttribBinding(VAO,0,0);
    glVertexArrayAttribBinding(VAO,1,0);
    glVertexArrayAttribBinding(VAO,2,0);

    Logs::debug("Mesh created with VBO: " + std::to_string(VBO) + ", EBO: " + std::to_string(EBO) + ", VAO: " + std::to_string(VAO));
}

void Mesh::draw(const Shader& shader) {     // render the mesh
    // bind appropriate textures
    unsigned int diffuseNr  = 1;
    unsigned int specularNr = 1;
    unsigned int normalNr   = 1;
    unsigned int heightNr   = 1;
    for(unsigned int i = 0; i < textures.size(); i++)
    {
        glActiveTexture(textures[i].code); // active proper texture unit before binding
        // retrieve texture number (the N in diffuse_textureN)
        std::string number;
        std::string name = textures[i].type;
        if(name == "texture_diffuse")
            number = std::to_string(diffuseNr++);
        else if(name == "texture_specular")
            number = std::to_string(specularNr++); // transfer unsigned int to string
        else if(name == "texture_normal")
            number = std::to_string(normalNr++); // transfer unsigned int to string
        else if(name == "texture_height")
            number = std::to_string(heightNr++); // transfer unsigned int to string

        // now set the sampler to the correct texture unit
        shader.setInt((name).c_str(), i);
        // and finally bind the texture
        glBindTexture(GL_TEXTURE_2D, textures[i].id);
    }

    // draw mesh
    glBindVertexArray(VAO);
    glDrawElementsBaseVertex(GL_TRIANGLES, (int) indices.size(), GL_UNSIGNED_INT, (void *) 0, 0);
    glBindVertexArray(0);


    // always good practice to set everything back to defaults once configured.
    glActiveTexture(GL_TEXTURE0);
}

int Mesh::loadTextures(const char* filename, unsigned int tCode,const std::string& name){
    TEXTURE texture;
    texture.type = name;
    texture.code = (int) tCode;
    glGenTextures(1, &texture.id);
    glActiveTexture(tCode);
    glBindTexture(GL_TEXTURE_2D, texture.id);
    Mesh::initTextures();
    // load and generate the texture
    int width, height, nrChannels;
    stbi_set_flip_vertically_on_load(1);
    unsigned char *data = stbi_load(filename, &width, &height, &nrChannels, 0);
    // printf("Loaded image, Width: %d Height: %d Channels: %d\n", width, height, nrChannels);
    if (data) {
        GLint format = GL_RGB;
        if (nrChannels == 4)
            format = GL_RGBA;
        else if (nrChannels == 1) {
            format = GL_RED;
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_R, GL_RED);
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_G, GL_RED);
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_B, GL_RED);
        }
        glTexImage2D(GL_TEXTURE_2D, 0, format, width, height, 0, format, GL_UNSIGNED_BYTE, data);
        glGenerateMipmap(GL_TEXTURE_2D);
        stbi_image_free(data);
    } else {
        printf("Failed to load texture\n");
        return -1;
    }
    this->textures.push_back(texture);
    return 0;
}

void Mesh::initTextures() {
    // set the texture wrapping/filtering options (on the currently bound texture object)
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
}

Mesh::~Mesh() {
    glDeleteVertexArrays(1, &VAO);
    glDeleteBuffers(1, &VBO);
    glDeleteBuffers(1, &EBO);
    for(auto & texture : textures){
        glDeleteTextures(1, &texture.id);
    }
}
